File: javascript/invertBinTree.js
/*
Problem 226: Invert Binary Tree

Invert a binary tree.

Example:
Input:
     4
   /   \
  2     7
 / \   / \
1   3 6   9
Output:
     4
   /   \
  7     2
 / \   / \
9   6 3   1

Trivia: This problem was inspired by this original tweet by Max Howell:
    Google: 90% of our engineers use the software you wrote (Homebrew), but you can’t invert
    a binary tree on a whiteboard so f*** off.
*/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
let invertTree = root => {
    if (!root) return null;
    
    let temp = invertTree(root.left);
    root.left = invertTree(root.right);
    root.right = temp;
    
    return root;
};

File: c/invertBinTree.c
/*
Problem 226: Invert Binary Tree

Invert a binary tree.

Example:
Input:
     4
   /   \
  2     7
 / \   / \
1   3 6   9
Output:
     4
   /   \
  7     2
 / \   / \
9   6 3   1

Trivia: This problem was inspired by this original tweet by Max Howell:
    Google: 90% of our engineers use the software you wrote (Homebrew), but you can’t invert
    a binary tree on a whiteboard so f*** off.

Solution memory usage: 5.4 MB, less than 86.11% of C submissions
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */


struct TreeNode* invertTree(struct TreeNode* root) {
    if (!root) return NULL;
    
    struct TreeNode* temp = invertTree(root->left);
    root->left = invertTree(root->right);
    root->right = temp;
    
    return root;
}
